triton-cli/src/main.rs

```diff
@@ -0,0 +1,210 @@
+) -> Result<()> {
+    if std::path::Path::new(out).exists() {
+        println!("output file already exists: {out}");
+        std::process::exit(1);
+    }
+    let asm = fs::read_to_string(asm_filepath)
+        .with_context(|| "Failed to read Triton assembly from file")?;
+    let program = triton_program!({ asm });
+
+    let public_input = PublicInput::from(parse_inputs(public_inputs));
+    let non_determinism = NonDeterminism::from(parse_inputs(private_inputs));
+    println!("proving...");
+    let data = triton_vm::prove_program(&program, public_input, non_determinism)
+        .with_context(|| "Triton VM errored during program execution")?;
+    println!("success!");
+
+    write_proof(data, out).with_context(|| "Failed to write proof to file")?;
+    println!("proof written to: {out}");
+    Ok(())
+}
+
+#[derive(Serialize, Deserialize)]
+struct SerializedProof {
+    version: u8,
+    security_level: u64,
+    fri_expansion_factor: u64,
+    num_trace_randomizers: u64,
+    num_colinearity_checks: u64,
+    num_combination_codeword_checks: u64,
+    program_digest: Vec<u64>,
+    public_inputs: Vec<u64>,
+    public_outputs: Vec<u64>,
+    proof: Vec<u64>,
+}
+
+fn read_proof(proof_path: &str) -> Result<(Stark, Claim, Proof)> {
+    let mut file = fs::File::open(proof_path)?;
+    let mut proof_bytes: Vec<u8> = Vec::new();
+    file.read_to_end(&mut proof_bytes)?;
+    let serialized_proof: SerializedProof = bincode::deserialize(&proof_bytes)?;
+    let stark = Stark {
+        security_level: usize::try_from(serialized_proof.security_level)?,
+        fri_expansion_factor: usize::try_from(serialized_proof.fri_expansion_factor)?,
+        num_trace_randomizers: usize::try_from(serialized_proof.num_trace_randomizers)?,
+        num_collinearity_checks: usize::try_from(serialized_proof.num_colinearity_checks)?,
+        num_combination_codeword_checks: usize::try_from(
+            serialized_proof.num_combination_codeword_checks,
+        )?,
+    };
+    let mut digest = Digest::default();
+    assert_eq!(
+        digest.0.len(),
+        serialized_proof.program_digest.len(),
+        "digest length mismatch!"
+    );
+    serialized_proof
+        .program_digest
+        .iter()
+        .enumerate()
+        .for_each(|(i, x)| digest.0[i] = BFieldElement::new(*x));
+    let claim = Claim {
+        program_digest: digest,
+        input: serialized_proof
+            .public_inputs
+            .iter()
+            .map(|v| BFieldElement::new(*v))
+            .collect(),
+        output: serialized_proof
+            .public_outputs
+            .iter()
+            .map(|v| BFieldElement::new(*v))
+            .collect(),
+    };
+    let proof_vec = serialized_proof
+        .proof
+        .iter()
+        .map(|v| BFieldElement::new(*v))
+        .collect();
+    Ok((stark, claim, Proof(proof_vec)))
+}
+
+fn write_proof(data: (Stark, Claim, Proof), out: &str) -> Result<()> {
+    let (stark, claim, proof) = data;
```

Comment:
If you want, you can pattern match the input directly:

```suggestion
fn write_proof((stark, claim, proof): (Stark, Claim, Proof), out: &str) -> Result<()> {
```

It's a matter of taste, though, and any decision here is entirely yours. :relieved:

Reply:
ah, didn't know i could do js style destructures 👍🏻

Reply:
I was happily surprised when I found out. :smiley: I mostly wanted to point out the possibility, please feel free to use either style. :relieved: